### ISSUE TITLE
Asymmetric getter/setter with/without None triggers erroneous 'is not a known attribute of "None"' error

### ISSUE THREAD

**Issue by gh-andre** [microsoft/pyright#10137]
**Describe the bug**

When a `@property` getter and setter are defined such that the getter returns `T|None` and the setter only allows to set a value of `T`, an error _is not a known attribute of "None_ is reported when trying to access that property right after it was assigned a non-`None` value.

**Code or Screenshots**

```python
class S:
    def __init__(self) -> None:
        self.a: int = 0
        
class E:
    def __init__(self) -> None:
        self.s: S = S()

class T:
    def __init__(self) -> None:
        self._e: E|None = None

    @property
    def e(self) -> E|None:
        return self._e

    @e.setter
    def e(self, v: E) -> None: # no error if E|None is used
        self._e = v

def fn(s: S) -> None:
    pass

t: T = T()
t.e = E()
fn(t.e.s) # ERROR: "s" is not a known attribute of "None" 
```

**VS Code extension or command-line**

I encountered this error in Visual Studio, but I can reproduce it in Playground with the latest Pyright version (1.1.397).

**Comment by erictraut** [comment]
Pyright is working as intended here, so I don't consider this a bug.

When pyright encounters an asymmetric descriptor (including a property), it does not perform type narrowing upon assignment. This behavior was discussed at length in public typing forums several years ago, and the consensus was that static type checkers should treat asymmetric descriptors specially and should not assume that the getter returns the same type as the setter. This hasn't yet been codified in the Python typing spec, but it's on my (long) list of things to add. For historical reference, refer to [this thread](https://github.com/microsoft/pyright/issues/1814) and [this thread](https://github.com/python/mypy/issues/3004).

**Comment by gh-andre** [comment]
Thanks for the links. One material difference between this case and those described in the linked threads is that in all those examples, there is some value conversion taking place, such as `str`-to-`int`, inheritance applied, etc, and in this case it's a union of types.

I would agree that when value conversion is performed, all bets are off and the getter cannot be assumed to return what was set, but in case of a union of types, the assumption of getter returning what was set is already made when getter/setter have the same type and no error is reported, even if the code is messed up like this:
```python
    ...
    @property
    def e(self) -> E|None:
        return self._e

    @e.setter
    def e(self, v: E|None) -> None:
        self._e = None

t: T = T()
t.e = E()
fn(t.e.s)  # no static typing error
```
It seems to me that unions of types should be treated differently than setter argument value conversions.

**Comment by gh-andre** [comment]
> A type checker has no way to know whether conversions are taking place or not

I didn't communicate my point well. I meant that the current logic accepted by the typing community is that with the same union of types in getters/setters, the getter will return what was set in the setter - that's how it currently works. I was trying to say that a subset of the original getter union of types in a setter would use the exact same logic, just with fewer types, so the implementer would be expected to assign appropriately in both cases, as it is today. No need to know whether any conversion is taking place or not. Works in TypeScript, for example.

> they set the wrong expectations for readers of the code.

Method calls won't work for narrowing and if one wants a mechanism in which some property can cannot be reset to `None`, the pattern I described is the most concise way to express this. Adding `None` to the union in such case would actually makes code more verbose and less readable.

Anyway, thanks for your insights, as always. Since this behavior has been finalized, I'll leave it at that.

**Comment by gh-andre** [comment]
... and just for completeness, there's no even a way to work this around today, as the usual methods don't work in this context, like this (same with `assert`):
```python
if t.e is not None:
    fn(t.e.s)   # still reports that t.e can be None
```

**Comment by erictraut** [comment]
Here's a workaround that is provably safe.

```python
if (e := t.e) is not None:
    fn(e.s)
```

**Comment by gh-andre** [comment]
Thanks. I realize that a new variable solves this. Just wanted to highlight that this behavior extends beyond the adjacent setting/evaluating logic shown in the OP and affects the getter too in its own context.
